src/node_large_page.cc

```diff
@@ -0,0 +1,253 @@
+// THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES
+// OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+// ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE
+// OR OTHER DEALINGS IN THE SOFTWARE.
+//
+// SPDX-License-Identifier: MIT
+
+#include <errno.h>
+#include <fcntl.h>  // _O_RDWR
+#include <limits.h>  // PATH_MAX
+#include <locale.h>
+#include <signal.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdint.h>
+#include <string.h>
+#include <sys/types.h>
+#include <sys/mman.h>
+
+#include <string>
+#include <fstream>
+#include <iostream>
+#include <sstream>
+
+// The functions in this file map the text segment of node into 2M pages.
+// The algorithm is simple
+// 1. Find the text region of node binary in memory
+//    Examine the /proc/self/maps to determine the currently mapped text
+//    region and obtain the start and end
+//    Modify the start to point to the very beginning of node text segment
+//    (from variable nodetext setup in ld.script)
+//    Align the address of start and end to Large Page Boundaries
+//
+// 2. Move the text region to large pages
+//    Map a new area and copy the original code there
+//    Use mmap using the start address with MAP_FIXED so we get exactly the
+//      same virtual address
+//    Use madvise with MADV_HUGE_PAGE to use Anonymous 2M Pages
+//    If successful copy the code there and unmap the original region.
+
+extern char __executable_start;
+extern char __etext;
+extern char __nodetext;
+
+namespace node {
+#define ALIGN(x, a)           (((x) + (a) - 1) & ~((a) - 1))
+#define PAGE_ALIGN_UP(x, a)   ALIGN(x, a)
+#define PAGE_ALIGN_DOWN(x, a) ((x) & ~((a) - 1))
+
+struct text_region {
+  char* from;
+  char* to;
+  int   total_hugepages;
+  bool  found_text_region;
+};
+
+static void PrintSystemError(int error) {
+  fprintf(stderr, "Hugepages WARNING: %s\n", strerror(error));
+  return;
+}
+
+//  The format of the maps file is the following
+//  address           perms offset  dev   inode       pathname
+//  00400000-00452000 r-xp 00000000 08:02 173521      /usr/bin/dbus-daemon
+
+static struct text_region FindNodeTextRegion() {
+  std::ifstream ifs;
+  std::string map_line;
+  std::string permission;
+  char dash;
+  int64_t  start, end;
+  const size_t hps = 2L * 1024 * 1024;
+  struct text_region nregion;
+
+  nregion.found_text_region = false;
+
+  ifs.open("/proc/self/maps");
+  if (!ifs) {
+    fprintf(stderr, "Could not open /proc/self/maps\n");
+    return nregion;
+  }
+  std::getline(ifs, map_line);
+  std::istringstream iss(map_line);
+  ifs.close();
+
+  iss >> std::hex >> start;
+  iss >> dash;
+  iss >> std::hex >> end;
+  iss >> permission;
+
+  if (permission.compare("r-xp") == 0) {
+    start = reinterpret_cast<unsigned int64_t>(&__nodetext);
+    char* from = reinterpret_cast<char *>PAGE_ALIGN_UP(start, hps);
+    char* to = reinterpret_cast<char *>PAGE_ALIGN_DOWN(end, hps);
```

Comment:
style nit: left-leaning pointers (e.g. `char*`), and I would not rely on the implicit parentheses added by the macros here

Reply:
Thanks. Fixed.